### PR TITLE
Pickle exceptions

### PR DESCRIPTION
**Pull request by joerunde**
**What this PR does / why we need it**:

This PR adds an exception pickler that will try its best to accommodate exceptions that do not properly pickle themselves. Unfortunately, these are common in the wild. This will also take care of serializing a full cause chain of exceptions, so that useful tracebacks can be generated by unpicled exceptions.

This is all useful in the context of a `DestroyableProcess`, which is a utility used by things like the `LocalModelTrainer` to delegate work to a subprocess. These subprocesses have to pickle errors back to the parent process.

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: caikit/core/concurrency/destroyable_process.py [hunk 1]
```diff
@@ -31,8 +31,9 @@
 import alog
 
 # Local
-from ..exceptions import error_handler
 from .destroyable import Destroyable
+from .pickling_exception import ExceptionPickler
+from caikit.core.exceptions import error_handler
```
Comment (joerunde) [inline caikit/core/concurrency/destroyable_process.py:36]: :hmmmmmm: no idea why this got switched around

File: caikit/core/concurrency/pickling_exception.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            __reduce__
+        """
+        state_dict = {
+            "exception": self.exception,
+            "cause": ExceptionPickler(self.exception.__cause__)
+            if self.exception.__cause__
+            else None,
+            "context": ExceptionPickler(self.exception.__context__)
+            if self.exception.__context__
+            else None,
```
Comment (mynhardtburger) [inline caikit/core/concurrency/pickling_exception.py:101]: NIT: Wrap in parentheses for readability

```suggestion
            "cause": (
                ExceptionPickler(self.exception.__cause__)
                if self.exception.__cause__
                else None
            ),
            "context": (
                ExceptionPickler(self.exception.__context__)
                if self.exception.__context__
                else None
            ),
```

**Review by mynhardtburger — APPROVED** [review]
lgtm :)

File: tests/core/concurrency/__init__.py
Comment (gabe-l-hart) [inline tests/core/concurrency/__init__.py:1]: Copyright header

**Review by gabe-l-hart — COMMENTED** [review]
One tiny NIT. I also have a medium-strong preference to put `concurrency` under `core.toolkit`, but otherwise it looks great!

File: caikit/core/concurrency/pickling_exception.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    def __setstate__(self, state_dict):
+        """Reconstructs the exception out of the state_dict that is returned by __getstate__"""
+        if "exception" in state_dict:
+            self.exception = state_dict["exception"]
+        else:
+            initializer = state_dict["initializer"]
+            self.exception = initializer(*state_dict["args"], **state_dict["kwargs"])
+
+        if state_dict["cause"]:
+            self.exception.__cause__ = state_dict["cause"].get()
```
[the remark this answers is not in the document]
Comment (joerunde) [reply caikit/core/concurrency/pickling_exception.py:82]: Yeah, I figured exceptions are a thing that have been this way forever and will likely continue to be this way, at least until a 4.x python comes along

File: caikit/core/concurrency/pickling_exception.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                else None
+            ),
+            "context": (
+                ExceptionPickler(self.exception.__context__)
+                if self.exception.__context__
+                else None
+            ),
+        }
+
+        # ty/catch pickle errors
```
[the remark this answers is not in the document]
Comment (joerunde) [reply caikit/core/concurrency/pickling_exception.py:108]: `ty` :D

File: caikit/core/concurrency/pickling_exception.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+                    # check that we can re-build this exception
+                    _ = state_dict["initializer"](
+                        *state_dict["args"], **state_dict["kwargs"]
+                    )
+                    log.debug4(
+                        "Successfully found all the args to re-initialize exception"
+                    )
+
+                except Exception as e:
```
[the remark this answers is not in the document]
Comment (joerunde) [reply caikit/core/concurrency/pickling_exception.py]: Huh, maybe `ruff` isn't configured to catch these?

**Review by evaline-ju — APPROVED** [review]
LGTM!

**Review by gabe-l-hart — APPROVED** [review]
Ship it!

**Pull request merged**